ql/src/java/org/apache/hadoop/hive/ql/metadata/Hive.java

```diff
@@ -1527,8 +1527,23 @@ public void truncateTable(String dbDotTableName, Map<String, String> partSpec, L
       }
 
       // TODO: APIs with catalog names
-      List<String> partNames = ((null == partSpec)
-        ? null : getPartitionNames(table.getDbName(), table.getTableName(), partSpec, (short) -1));
+      List<String> partNames = null;
+      if (table.getStorageHandler() != null && table.getStorageHandler().alwaysUnpartitioned() && partSpec != null) {
```

Comment:
can we have a method in StorageHandler getPartitions()?

Reply:
Added a getPartitionsByFilter() API.